### PR TITLE
build: update project tooling

### PR DESCRIPTION
**Pull request by NoNameProvided**
This PR: 
  - updates .gitignore file to exclude common OS generated files
  - updates the year in the license
  - configures TS and Jest with more modern settings
  - adds Prettier (formatting has not been done as it would cause conflict with every PR)
  - adds Husky and lint-staged for auto-formatting during commits (new code additions become well formatted)
  - updates CI configs to 
     - run lint, format, test, build in different tasks for faster feedback
     - removes Travis in favor of GH Actions
     - adds support for auto-deploying GH releases
 

In a separate PR I also need to add some documentation about:
  - how make a release
  - commit and PR naming schemes
  - some other possible tooling information

### CODE REVIEW

File: .prettierrc.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+printWidth: 120
+tabWidth: 2
+useTabs: false
+semi: true
+singleQuote: true
+trailingComma: es5
+bracketSpacing: true
+arrowParens: avoid
```
Comment (saulotoledo) [inline .prettierrc.yml:8]: I am in favour of having a newline in the end for all files. I know this is a personal opinion, but by default we have all `.ts` files containing EOF newlines, so I believe we could extend for all files then.

File: codecov.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (saulotoledo) [inline codecov.yml:1]: This line may be removed.

File: package-lock.json [hunk 3]
```diff
@@ -803,6 +803,12 @@
       "integrity": "sha512-f5j5b/Gf71L+dbqxIpQ4Z2WlmI/mPJ0fOkGGmFgtb6sAu97EPczzbS3/tJKxmcYDj55OX6ssqwDAWOHIYDRDGA==",
       "dev": true
     },
+    "@types/parse-json": {
```
Comment (saulotoledo) [inline package-lock.json:806]: To discuss: what do you think about taking the chance and replace npm with Yarn?

File: package.json [hunk 4]
```diff
@@ -43,22 +56,25 @@
...
-    "lint": "eslint --config ./.eslintrc.js --ext .ts ./src ./test",
-    "package": "npm run build && npm run copy && npm run public && rimraf build/compiled",
-    "pretest": "npm run lint",
-    "public": "json -I -f build/package/package.json -e 'this.private=false'",
-    "test": "jest --coverage"
+    "build": "rimraf build && tsc --project tsconfig.prod.json",
+    "prettier:fix": "prettier --write \"**/*.{ts,md}\"",
+    "prettier:check": "prettier --check \"**/*.{ts,md}\"",
+    "lint:fix": "eslint --max-warnings 0 --fix --ext .ts src/",
+    "lint:check": "eslint --max-warnings 0 --ext .ts src/",
```
[the remark this answers is not in the document]
Comment (NoNameProvided) [reply package.json]: We have a command for autofix in the above line: "lint:fix": "eslint --max-warnings 0 --fix --ext .ts src/",`

File: .prettierrc.yml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+printWidth: 120
+tabWidth: 2
```
[the remark this answers is not in the document]
Comment (NoNameProvided) [reply .prettierrc.yml:2]: Yes, I know. I did not run prettier for this reason. We should clean up the PRs and only autoformat after.

File: .prettierrc.yml (hunk 1 above)
Comment (NoNameProvided) [reply .prettierrc.yml]: Fixed, I have added a new line to all of the config files which lacked it.

File: package-lock.json (hunk 3 above)
Comment (NoNameProvided) [reply package-lock.json:1197]: I strongly prefer NPM over Yarn. The modern versions of NPM rendered the benefits of Yarn to zero and also you cannot publish to npmjs.com with Yarn so we would need to use NPM anyway (just in the CI, but still two tool).

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (saulotoledo) [reply package.json]: Ops, I mean `"lint:fix": "npm lint:check -- --fix",`

File: package-lock.json (hunk 3 above)
Comment (saulotoledo) [reply package-lock.json:1197]: ok :)

**Comment by saulotoledo** [comment]
LGTM.

**Pull request merged**